library/mps/reader.c

```diff
@@ -0,0 +1,490 @@
+#include "common.h"
+#include "trace.h"
+
+#include <string.h>
+
+#if defined(MBEDTLS_MPS_TRACE)
+static int trace_id = TRACE_BIT_READER;
+#endif /* MBEDTLS_MPS_TRACE */
+
+/*
+ * GENERAL NOTE ON CODING STYLE
+ *
+ * The following code intentionally separates memory loads
+ * and stores from other operations (arithmetic or branches).
+ * This leads to the introduction of many local variables
+ * and significantly increases the C-code line count, but
+ * should not increase the size of generated assembly.
+ *
+ * This reason for this is twofold:
+ * (1) It will ease verification efforts using the VST
+ *     whose program logic cannot directly reason
+ *     about instructions containing a load or store in
+ *     addition to other operations (e.g. *p = *q or
+ *     tmp = *p + 42).
+ * (2) Operating on local variables and writing the results
+ *     back to the target contexts on success only
+ *     allows to maintain structure invariants even
+ *     on failure - this in turn has two benefits:
+ *     (2.a) If for some reason an error code is not caught
+ *           and operation continues, functions are nonetheless
+ *           called with sane contexts, reducing the risk
+ *           of dangerous behavior.
+ *     (2.b) Randomized testing is easier if structures
+ *           remain intact even in the face of failing
+ *           and/or non-sensical calls.
+ *     Moreover, it might even reduce code-size because
+ *     the compiler need not write back temporary results
+ *     to memory in case of failure.
+ *
+ */
+
+int mbedtls_reader_init( mbedtls_reader *rd,
+                         unsigned char *acc,
+                         mbedtls_mps_size_t acc_len )
+{
+    mbedtls_reader const zero = { NULL, 0, 0, 0, 0, NULL, 0, 0, { 0 } };
+    *rd = zero;
+    TRACE_INIT( "reader_init, acc len %u", (unsigned) acc_len );
+
+    rd->acc = acc;
+    rd->acc_len = acc_len;
+    RETURN( 0 );
+}
+
+int mbedtls_reader_free( mbedtls_reader *rd )
+{
+    mbedtls_reader const zero = { NULL, 0, 0, 0, 0, NULL, 0, 0, { 0 } };
+    TRACE_INIT( "reader_free" );
+    *rd = zero;
+    /* TODO: Use reliable way of zeroization, provided
+     * zeroization is necessary in the first place. If so,
+     * consider also zeroizing the accumulator buffer. */
+    RETURN( 0 );
+}
+
+int mbedtls_reader_feed( mbedtls_reader *rd,
+                         unsigned char *new_frag,
+                         mbedtls_mps_size_t new_frag_len )
+{
+    unsigned char *acc;
+    mbedtls_mps_size_t copy_to_acc;
+    TRACE_INIT( "reader_feed, frag %p, len %u",
+                (void*) new_frag, (unsigned) new_frag_len );
+
+    if( new_frag == NULL )
+        RETURN( MBEDTLS_ERR_MPS_READER_INVALID_ARG );
+
+    MBEDTLS_MPS_STATE_VALIDATE_RAW( rd->frag == NULL,
+        "mbedtls_reader_feed() requires reader to be in producing mode" );
+
+    acc = rd->acc;
+    if( acc != NULL )
+    {
+        mbedtls_mps_size_t aa, ar;
+
+        ar = rd->acc_share.acc_remaining;
```

Comment:
`acc_avail` and `acc_remaining` are both `0` on init, that's right. They're only non-zero in case we're currently using the accumulator to gather enough data to fulfill a previous request to `mbedtls_mps_reader_get()`.

Reply:
See also the documentation:

Reply:
The flow is the following:
a) Initially, the accumulator is not used, and on the first `mbedtls_mps_reader_feed()` by the 'producer' (say, record layer), control immediately passes to the consumer (say, handshake layer).
b) When the consumer asks for more than what's available in the current fragment via `mbedtls_mps_reader_get()`, the reader will remember that, and configure `acc_remaining` so that subsequent `mbedtls_mps_reader_feed()` calls only succeed once `acc_remaining` bytes have been gathered.